// @react-pdf-viewer/print/lib/cjs/print.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@react-pdf-viewer/core');
var React = require('react');
var reactDom = require('react-dom');

function _interopNamespace(e) {
    if (e && e.__esModule) return e;
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n["default"] = e;
    return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

var PrintIcon = function () { return (React__namespace.createElement(core.Icon, { size: 16 },
    React__namespace.createElement("path", { d: "M7.5,19.499h9 M7.5,16.499h9 M5.5,16.5h-3c-1.103-0.003-1.997-0.897-2-2v-6c0.003-1.103,0.897-1.997,2-2h19\n            c1.103,0.003,1.997,0.897,2,2v6c-0.003,1.103-0.897,1.997-2,2h-3\n            M5.5,4.5v-4h9.586c0.265,0,0.52,0.105,0.707,0.293l2.414,2.414\n            C18.395,3.394,18.5,3.649,18.5,3.914V4.5\n            M18.5,22.5c0,0.552-0.448,1-1,1h-11c-0.552,0-1-0.448-1-1v-9h13V22.5z\n            M3.5,8.499\n            c0.552,0,1,0.448,1,1s-0.448,1-1,1s-1-0.448-1-1S2.948,8.499,3.5,8.499z\n            M14.5,0.499v4h4" }))); };

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

var TOOLTIP_OFFSET = { left: 0, top: 8 };
var PrintButton = function (_a) {
    var enableShortcuts = _a.enableShortcuts, onClick = _a.onClick;
    var l10n = React__namespace.useContext(core.LocalizationContext).l10n;
    var label = l10n && l10n.print ? l10n.print.print : 'Print';
    var ariaKeyShortcuts = enableShortcuts ? (core.isMac() ? 'Meta+P' : 'Ctrl+P') : '';
    return (React__namespace.createElement(core.Tooltip, { ariaControlsSuffix: "print", position: core.Position.BottomCenter, target: React__namespace.createElement(core.MinimalButton, { ariaKeyShortcuts: ariaKeyShortcuts, ariaLabel: label, testId: "print__button", onClick: onClick },
            React__namespace.createElement(PrintIcon, null)), content: function () { return label; }, offset: TOOLTIP_OFFSET }));
};

var PrintStatus;
(function (PrintStatus) {
    PrintStatus["CheckingPermission"] = "CheckingPermission";
    PrintStatus["Inactive"] = "Inactive";
    PrintStatus["Preparing"] = "Preparing";
    PrintStatus["Cancelled"] = "Cancelled";
    PrintStatus["Ready"] = "Ready";
})(PrintStatus || (PrintStatus = {}));

var Print = function (_a) {
    var children = _a.children, enableShortcuts = _a.enableShortcuts, store = _a.store;
    var print = function () {
        store.update('printStatus', PrintStatus.CheckingPermission);
    };
    var render = children || PrintButton;
    return render({
        enableShortcuts: enableShortcuts,
        onClick: print,
    });
};

var PERMISSION_PRINT = 4;
var PERMISSION_PRINT_HIGHT_QUALITY = 2048;
var CheckPrintPermission = function (_a) {
    var doc = _a.doc, store = _a.store;
    var l10n = React__namespace.useContext(core.LocalizationContext).l10n;
    var _b = React__namespace.useState(true), isAllowed = _b[0], setIsAllowed = _b[1];
    React__namespace.useEffect(function () {
        doc.getPermissions().then(function (permissions) {
            var canPrint = permissions === null ||
                permissions.includes(PERMISSION_PRINT) ||
                permissions.includes(PERMISSION_PRINT_HIGHT_QUALITY);
            canPrint ? store.update('printStatus', PrintStatus.Preparing) : setIsAllowed(false);
        });
    }, []);
    return isAllowed ? (React__namespace.createElement(React__namespace.Fragment, null)) : (React__namespace.createElement(core.Modal, { ariaControlsSuffix: "print-permission", closeOnClickOutside: false, closeOnEscape: false, content: function (toggle) {
            var close = function () {
                toggle();
                store.update('printStatus', PrintStatus.Cancelled);
            };
            return (React__namespace.createElement(React__namespace.Fragment, null,
                React__namespace.createElement("div", { className: "rpv-print__permission-body" }, l10n && l10n.print
                    ? l10n.print.disallowPrint
                    : 'The document does not allow to print'),
                React__namespace.createElement("div", { className: "rpv-print__permission-footer" },
                    React__namespace.createElement(core.Button, { onClick: close }, l10n && l10n.print ? l10n.print.close : 'Close'))));
        }, isOpened: true }));
};

var PrintProgress = function (_a) {
    var numLoadedPages = _a.numLoadedPages, numPages = _a.numPages, onCancel = _a.onCancel;
    var l10n = React__namespace.useContext(core.LocalizationContext).l10n;
    var direction = React__namespace.useContext(core.ThemeContext).direction;
    var isRtl = direction === core.TextDirection.RightToLeft;
    var progress = Math.floor((numLoadedPages * 100) / numPages);
    return (React__namespace.createElement("div", { className: "rpv-print__progress" },
        React__namespace.createElement("div", { className: core.classNames({
                'rpv-print__progress-body': true,
                'rpv-print__progress-body--rtl': isRtl,
            }) },
            React__namespace.createElement("div", { className: "rpv-print__progress-message" }, l10n && l10n.print
                ? l10n.print.preparingDocument
                : 'Preparing document ...'),
            React__namespace.createElement("div", { className: "rpv-print__progress-bar" },
                React__namespace.createElement(core.ProgressBar, { progress: progress })),
            React__namespace.createElement(core.Button, { onClick: onCancel }, l10n && l10n.print ? l10n.print.cancel : 'Cancel'))));
};

var PageThumbnail = function (_a) {
    var canvas = _a.canvas, page = _a.page, pageHeight = _a.pageHeight, pageWidth = _a.pageWidth, rotation = _a.rotation, onLoad = _a.onLoad;
    var renderTask = React__namespace.useRef();
    var _b = React__namespace.useState(''), src = _b[0], setSrc = _b[1];
    React__namespace.useEffect(function () {
        var task = renderTask.current;
        if (task) {
            task.cancel();
        }
        var printUnit = 150 / 72;
        canvas.height = Math.floor(pageHeight * printUnit);
        canvas.width = Math.floor(pageWidth * printUnit);
        var canvasContext = canvas.getContext('2d');
        canvasContext.save();
        canvasContext.fillStyle = 'rgb(255, 255, 255)';
        canvasContext.fillRect(0, 0, canvas.width, canvas.height);
        canvasContext.restore();
        var viewport = page.getViewport({ rotation: rotation, scale: 1 });
        renderTask.current = page.render({
            canvasContext: canvasContext,
            intent: 'print',
            transform: [printUnit, 0, 0, printUnit, 0, 0],
            viewport: viewport,
        });
        renderTask.current.promise.then(function () {
            'toBlob' in canvas
                ? canvas.toBlob(function (blob) {
                    setSrc(URL.createObjectURL(blob));
                })
                : setSrc(canvas.toDataURL());
        }, function () {
        });
    }, []);
    return (src && (React__namespace.createElement("div", { className: "rpv-print__page" },
        React__namespace.createElement("img", { src: src, onLoad: function () { return onLoad(); } }))));
};

var PageThumbnailContainer = function (_a) {
    var canvas = _a.canvas, doc = _a.doc, pageHeight = _a.pageHeight, pageIndex = _a.pageIndex, pageRotation = _a.pageRotation, pageWidth = _a.pageWidth, rotation = _a.rotation, onLoad = _a.onLoad;
    var _b = React__namespace.useState({
        height: pageHeight,
        page: null,
        viewportRotation: 0,
        width: pageWidth,
    }), pageSize = _b[0], setPageSize = _b[1];
    var page = pageSize.page, height = pageSize.height, width = pageSize.width;
    var isVertical = Math.abs(rotation + pageRotation) % 180 === 0;
    React__namespace.useEffect(function () {
        core.getPage(doc, pageIndex).then(function (pdfPage) {
            var viewport = pdfPage.getViewport({ scale: 1 });
            setPageSize({
                height: viewport.height,
                page: pdfPage,
                viewportRotation: viewport.rotation,
                width: viewport.width,
            });
        });
    }, []);
    var rotationNumber = (pageSize.viewportRotation + rotation + pageRotation) % 360;
    return (page && (React__namespace.createElement(PageThumbnail, { canvas: canvas, page: page, pageHeight: isVertical ? height : width, pageWidth: isVertical ? width : height, rotation: rotationNumber, onLoad: onLoad })));
};

var PrintZone = function (_a) {
    var doc = _a.doc, numLoadedPages = _a.numLoadedPages, pagesRotation = _a.pagesRotation, pageHeight = _a.pageHeight, pageWidth = _a.pageWidth, printStatus = _a.printStatus, rotation = _a.rotation, onCancel = _a.onCancel, onLoad = _a.onLoad;
    var canvas = React__namespace.useMemo(function () { return document.createElement('canvas'); }, []);
    var container = React__namespace.useMemo(function () {
        var zoneEle = document.querySelector('.rpv-print__zone');
        if (zoneEle) {
            return zoneEle;
        }
        var div = document.createElement('div');
        div.classList.add('rpv-print__zone');
        document.body.appendChild(div);
        return div;
    }, []);
    React__namespace.useEffect(function () {
        if (printStatus === PrintStatus.Ready) {
            document.documentElement.classList.add('rpv-print__html-printing');
            document.body.classList.add('rpv-print__body-printing');
            window.print();
        }
        var handler = function () {
            if (printStatus === PrintStatus.Ready) {
                document.documentElement.classList.remove('rpv-print__html-printing');
                document.body.classList.remove('rpv-print__body-printing');
                var zones = document.querySelectorAll('.rpv-print__zone');
                if (zones) {
                    zones.forEach(function (zoneEle) {
                        zoneEle.parentElement.removeChild(zoneEle);
                    });
                }
                canvas.height = 0;
                canvas.width = 0;
                document.removeEventListener('mousemove', handler);
                onCancel();
            }
        };
        document.addEventListener('mousemove', handler);
        return function () { return document.removeEventListener('mousemove', handler); };
    }, [printStatus]);
    return reactDom.createPortal(React__namespace.createElement(React__namespace.Fragment, null,
        Array(Math.min(numLoadedPages + 1, doc.numPages))
            .fill(0)
            .map(function (_, index) { return (React__namespace.createElement(PageThumbnailContainer, { key: index, canvas: canvas, doc: doc, pageHeight: pageHeight, pageIndex: index, pageRotation: pagesRotation.has(index) ? pagesRotation.get(index) : 0, pageWidth: pageWidth, rotation: rotation, onLoad: onLoad })); }),
        React__namespace.createElement("style", { dangerouslySetInnerHTML: {
                __html: "@page { size: ".concat(pageWidth, "pt ").concat(pageHeight, "pt }"),
            } })), container);
};

var PrintContainer = function (_a) {
    var doc = _a.doc, pagesRotation = _a.pagesRotation, pageHeight = _a.pageHeight, pageWidth = _a.pageWidth, rotation = _a.rotation, store = _a.store;
    var _b = React__namespace.useState(PrintStatus.Inactive), printStatus = _b[0], setPrintStatus = _b[1];
    var _c = React__namespace.useState(0), numLoadedPagesForPrint = _c[0], setNumLoadedPagesForPrint = _c[1];
    var cancelPrinting = function () {
        setNumLoadedPagesForPrint(0);
        setPrintStatus(PrintStatus.Inactive);
    };
    var handlePrintStatus = function (status) { return setPrintStatus(status); };
    var onLoadPage = function () {
        var total = numLoadedPagesForPrint + 1;
        if (total <= doc.numPages) {
            setNumLoadedPagesForPrint(total);
            total === doc.numPages && setPrintStatus(PrintStatus.Ready);
        }
    };
    React__namespace.useEffect(function () {
        store.subscribe('printStatus', handlePrintStatus);
        return function () {
            store.unsubscribe('printStatus', handlePrintStatus);
        };
    }, []);
    return (React__namespace.createElement(React__namespace.Fragment, null,
        printStatus === PrintStatus.CheckingPermission && React__namespace.createElement(CheckPrintPermission, { doc: doc, store: store }),
        printStatus === PrintStatus.Preparing && (React__namespace.createElement(PrintProgress, { numLoadedPages: numLoadedPagesForPrint, numPages: doc.numPages, onCancel: cancelPrinting })),
        (printStatus === PrintStatus.Preparing || printStatus === PrintStatus.Ready) &&
            numLoadedPagesForPrint <= doc.numPages && (React__namespace.createElement(PrintZone, { doc: doc, numLoadedPages: numLoadedPagesForPrint, pagesRotation: pagesRotation, pageHeight: pageHeight, pageWidth: pageWidth, printStatus: printStatus, rotation: rotation, onCancel: cancelPrinting, onLoad: onLoadPage }))));
};

var PrintMenuItem = function (_a) {
    var onClick = _a.onClick;
    var l10n = React__namespace.useContext(core.LocalizationContext).l10n;
    var label = l10n && l10n.print ? l10n.print.print : 'Print';
    return (React__namespace.createElement(core.MenuItem, { icon: React__namespace.createElement(PrintIcon, null), testId: "print__menu", onClick: onClick }, label));
};

var ShortcutHandler = function (_a) {
    var containerRef = _a.containerRef, store = _a.store;
    var keydownHandler = function (e) {
        if (e.shiftKey || e.altKey || e.key !== 'p') {
            return;
        }
        var isCommandPressed = core.isMac() ? e.metaKey : e.ctrlKey;
        if (!isCommandPressed) {
            return;
        }
        var containerEle = containerRef.current;
        if (!containerEle || !document.activeElement || !containerEle.contains(document.activeElement)) {
            return;
        }
        e.preventDefault();
        store.update('printStatus', PrintStatus.Preparing);
    };
    React__namespace.useEffect(function () {
        var containerEle = containerRef.current;
        if (!containerEle) {
            return;
        }
        document.addEventListener('keydown', keydownHandler);
        return function () {
            document.removeEventListener('keydown', keydownHandler);
        };
    }, [containerRef.current]);
    return React__namespace.createElement(React__namespace.Fragment, null);
};

var printPlugin = function (props) {
    var printPluginProps = React__namespace.useMemo(function () { return Object.assign({}, { enableShortcuts: true }, props); }, []);
    var store = React__namespace.useMemo(function () {
        return core.createStore({
            printStatus: PrintStatus.Inactive,
        });
    }, []);
    var PrintDecorator = function (props) { return (React__namespace.createElement(Print, __assign({ enableShortcuts: printPluginProps.enableShortcuts }, props, { store: store }))); };
    var PrintButtonDecorator = function () { return React__namespace.createElement(PrintDecorator, null, function (props) { return React__namespace.createElement(PrintButton, __assign({}, props)); }); };
    var PrintMenuItemDecorator = function (props) { return (React__namespace.createElement(PrintDecorator, null, function (p) { return (React__namespace.createElement(PrintMenuItem, { onClick: function () {
            p.onClick();
            props.onClick();
        } })); })); };
    var renderViewer = function (props) {
        var slot = props.slot;
        var updateSlot = {
            children: (React__namespace.createElement(React__namespace.Fragment, null,
                printPluginProps.enableShortcuts && (React__namespace.createElement(ShortcutHandler, { containerRef: props.containerRef, store: store })),
                React__namespace.createElement(PrintContainer, { doc: props.doc, pagesRotation: props.pagesRotation, pageHeight: props.pageHeight, pageWidth: props.pageWidth, rotation: props.rotation, store: store }),
                slot.children)),
        };
        return __assign(__assign({}, slot), updateSlot);
    };
    return {
        renderViewer: renderViewer,
        Print: PrintDecorator,
        PrintButton: PrintButtonDecorator,
        PrintMenuItem: PrintMenuItemDecorator,
    };
};

exports.PrintIcon = PrintIcon;
exports.printPlugin = printPlugin;
