// parameterize-js/helper/downcoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_DOWNCODER = exports.Downcoder = void 0;
const keymaps_1 = require("./keymaps");
class Downcoder {
    constructor(...keymaps) {
        const finalKeymap = {};
        const chars = [];
        for (const keymap of keymaps) {
            for (const key in keymap) {
                if (!keymap.hasOwnProperty(key)) {
                    continue;
                }
                finalKeymap[key] = keymap[key];
                chars.push(key);
            }
        }
        this.keymap = finalKeymap;
        this.chars = chars;
        this.regex = new RegExp('[' + this.chars.join() + ']|[^' + this.chars.join() + ']+', 'g');
    }
    apply(slug) {
        const pieces = slug.match(this.regex);
        if (pieces == null) {
            return slug;
        }
        return pieces.map(piece => {
            if (piece.length === 1) {
                const mapped = this.keymap[piece];
                return mapped || piece;
            }
            return piece;
        }).join();
    }
}
exports.Downcoder = Downcoder;
exports.DEFAULT_DOWNCODER = new Downcoder(keymaps_1.LATIN_MAP, keymaps_1.LATIN_SYMBOLS_MAP, keymaps_1.GREEK_MAP, keymaps_1.TURKISH_MAP, keymaps_1.ROMANIAN_MAP, keymaps_1.RUSSIAN_MAP, keymaps_1.UKRAINIAN_MAP, keymaps_1.CZECH_MAP, keymaps_1.POLISH_MAP, keymaps_1.LATVIAN_MAP, keymaps_1.ARABIC_MAP, keymaps_1.LITHUANIAN_MAP, keymaps_1.SERBIAN_MAP, keymaps_1.AZERBAIJANI_MAP, keymaps_1.GEORGIAN_MAP, keymaps_1.SLOVAK_MAP);
